integration/docker/csi/alluxio/nodeserver.go

```diff
@@ -87,45 +119,184 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
-	)
+	if !notMnt {
+		glog.V(4).Infoln("target path is already mounted")
+		return &csi.NodePublishVolumeResponse{}, nil
+	}
+
+	args := []string{"--bind", stagingPath, targetPath}
+	command := exec.Command("mount", args...)
+	glog.V(4).Infoln(command)
+	stdoutStderr, err := command.CombinedOutput()
+	glog.V(4).Infoln(string(stdoutStderr))
+	if err != nil {
+		if os.IsPermission(err) {
+			return nil, status.Error(codes.PermissionDenied, err.Error())
+		}
+		if strings.Contains(err.Error(), "invalid argument") {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &csi.NodePublishVolumeResponse{}, nil
+}
+
+func (ns *nodeServer) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpublishVolumeRequest) (*csi.NodeUnpublishVolumeResponse, error) {
+	targetPath := req.GetTargetPath()
+	command := exec.Command("/opt/alluxio/integration/fuse/bin/alluxio-fuse", "umount", targetPath)
+	glog.V(4).Infoln(command)
 	stdoutStderr, err := command.CombinedOutput()
 	if err != nil {
 		glog.V(3).Infoln(err)
 	}
 	glog.V(4).Infoln(string(stdoutStderr))
 
-	err = mount.CleanupMountPoint(req.GetTargetPath(), mount.New(""), false)
+	err = mount.CleanupMountPoint(targetPath, mount.New(""), false)
 	if err != nil {
 		glog.V(3).Infoln(err)
+	} else {
+		glog.V(4).Infof("Succeed in unmounting %s", targetPath)
 	}
-
 	return &csi.NodeUnpublishVolumeResponse{}, nil
 }
 
-func (ns *nodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstageVolumeRequest) (*csi.NodeUnstageVolumeResponse, error) {
-	return &csi.NodeUnstageVolumeResponse{}, nil
-}
-
 func (ns *nodeServer) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
+	if req.GetVolumeContext()["mountInPod"] == "true" {
+		ns.mutex.Lock()
+		defer ns.mutex.Unlock()
+
+		glog.V(4).Infoln("Creating Alluxio-fuse pod and mounting Alluxio to global mount point.")
+		fusePod, err := getAndCompleteFusePodObj(ns.nodeId, req)
+		if err != nil {
+			return nil, err
+		}
+		if _, err := ns.client.CoreV1().Pods(os.Getenv("NAMESPACE")).Create(fusePod); err != nil {
+			return nil, status.Errorf(codes.Internal, "Failed to launch Fuse Pod at %v.\n%v", ns.nodeId, err.Error())
+		}
+		glog.V(4).Infoln("Successfully creating Fuse pod.")
+
+		// Wait for alluxio-fuse pod finishing mount to global mount point
+		i := 0
+		for i < 12 {
+			time.Sleep(5 * time.Second)
+			command := exec.Command("bash", "-c", fmt.Sprintf("mount | grep %v | grep alluxio-fuse", req.GetStagingTargetPath()))
+			stdout, err := command.CombinedOutput()
+			if err != nil {
+				glog.V(3).Infoln("Alluxio is not mounted yet.")
+			}
+			if len(stdout) > 0 {
+				break
+			}
+			i++
+		}
+		if i == 12 {
+			glog.V(3).Infoln("alluxio-fuse is not mounted to global mount point in 60s.")
+			return nil, status.Error(codes.DeadlineExceeded, "alluxio-fuse is not mounted to global mount point in 60s")
+		}
+	}
```

Comment:
Style nit: Assume the exterior of the for-loop is the error case if you are waiting on a timeout with retries.

So at the top of the method you'd need:
```
if req.GetVolumeContext()["mountInPod"] == "false" {
    return &csi.NodeStageVolumeResponse{}, nil
}
```

And then this for-loop would look like the following:
```suggestion
		for i := 0; i < 12; i++ {
			time.Sleep(5 * time.Second)
			command := exec.Command("bash", "-c", fmt.Sprintf("mount | grep %v | grep alluxio-fuse", req.GetStagingTargetPath()))
			stdout, err := command.CombinedOutput()
			if err != nil {
				glog.V(3).Infoln("Alluxio is not mounted yet.")
			}
			if len(stdout) > 0 {
				return &csi.NodeStageVolumeResponse{}, nil
			}
		}
	}
	glog.V(3).Infoln("alluxio-fuse is not mounted to global mount point in 60s.")
	return nil, status.Error(codes.DeadlineExceeded, "alluxio-fuse is not mounted to global mount point in 60s")
```